### PR TITLE
fix: replace string_view with string in RaxTreeMap::FindIterator

### PR DESCRIPTION
**Pull request by kostasrim**
The main problem is that `find(std::string_view key)` returns a `FindIterator` which contains the `key` that was passed as argument in the call to `find` and not the actual `key contained and allocated within rax` and is easy dangle: 

```
  FindIterator res = find(std::string("some_key")); 
```
res now contains a dangling pointer because the internal string_view now points to the temporary (std::string("some_key")) which got deallocated at the end of the expression above.

The simplest solution is to make `FindIterator` own the key. (yes it does copy which is not great but it avoids dangling pointers). 

We can always optimize the copy later.

Also fixes a memory leak in the destructor: we need to iterate over all the available children in the rax tree and delete the data we allocated before we call `raxFree`.

* fixes dangling pointer caused by find()
* fixes memory leak in the destructor

### CODE REVIEW

File: src/core/search/rax_tree.h [hunk 1]
```diff
@@ -92,6 +93,11 @@ template <typename V> struct RaxTreeMap {
   }
 
   ~RaxTreeMap() {
+    for (auto it = begin(); it != end(); ++it) {
+      V* ptr = &(*it).second;
+      std::allocator_traits<decltype(alloc_)>::destroy(alloc_, ptr);
+      alloc_.deallocate(ptr, 1);
+    }
```
Comment (dranikpg) [inline src/core/search/rax_tree.h:100]: I really forgot this 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻 🤦🏻

File: src/core/search/rax_tree.h [hunk 2]
```diff
@@ -113,7 +119,8 @@ template <typename V> struct RaxTreeMap {
 
   FindIterator find(std::string_view key) const {
     if (void* ptr = raxFind(tree_, to_key_ptr(key), key.size()); ptr != raxNotFound)
-      return FindIterator{std::pair<std::string_view, V&>(key, *reinterpret_cast<V*>(ptr))};
+      return FindIterator{std::pair<std::string, V&>(std::string(key), *reinterpret_cast<V*>(ptr))};
```
Comment (dranikpg) [inline src/core/search/rax_tree.h:122]: I know it's dangerous, but it also means every lookup is a copy

**Review by dranikpg — APPROVED** [review]

File: src/core/search/rax_tree.h (hunk 2 above)
Comment (kostasrim) [reply src/core/search/rax_tree.h:122]: Well we already leaked in the tests :smile: I will figure out something and push it. I will merge this and follow up

**Pull request merged**